apps/data-pipeline/library-traffic-scraper/src/lib.ts

```diff
@@ -0,0 +1,201 @@
+
+  if (!scriptText) throw new Error("Could not find <script> containing locationIds");
+
+  const locMatch = scriptText.match(/let\s+locationIds\s*=\s*\[([\s\S]*?)]/);
+  const floorMatch = scriptText.match(/let\s+floors\s*=\s*\[([\s\S]*?)]/);
+  if (!locMatch || !floorMatch) throw new Error("Unable to capture locationIds or floors array");
+
+  const ids = parseArray(locMatch[1]);
+  const codes = parseArray(floorMatch[1]);
+
+  if (ids.length !== codes.length)
+    throw new Error("locationIds and floors arrays are different lengths");
+
+  const codeToLibrary = (code: string): string => {
+    if (code.includes("GSC")) return "Gateway Study Center";
+    if (code.includes("-SL-")) return "Science Library"; // SL prefix in list
+    return "Langson Library"; // fallback
+  };
+
+  const meta: Record<string, LocationMeta> = {};
+
+  ids.forEach((id, idx) => {
+    const code = codes[idx];
+    const root = $(`#leftSide${code}`);
+
+    const primary = root.find("h2.card-title").first().text().trim();
+
+    const sub = root.find("span.subLocation").first().text().trim().replace(/\s+/g, " ");
+
+    const label = sub ? `${primary} - ${sub}` : primary || "Unknown";
+
+    meta[id] = {
+      id,
+      floorCode: code,
+      libraryName: codeToLibrary(code),
+      locationLabel: label,
+    };
+  });
+
+  return meta;
+}
+
+async function fetchLocation(id: string): Promise<RawRespOK["data"] | null> {
+  const url = `https://www.lib.uci.edu/sites/all/scripts/occuspace.php?id=${id}`;
+  try {
+    const raw = await fetch(url).then((r) => r.text());
+    const parsedOnce = JSON.parse(raw);
+    const parsed = typeof parsedOnce === "string" ? JSON.parse(parsedOnce) : parsedOnce;
+    if (parsed.data && typeof parsed.data === "object") return (parsed as RawRespOK).data;
+    console.warn(`ID ${id} responded with error: ${(parsed as RawRespErr).error}`);
+  } catch (err) {
+    console.error(`Unexpected error while fetching ID ${id}:`, err);
+    throw err;
+  }
+  return null;
+}
+
+// - "active": Library location is open — perform scrape every 15 minutes.
+// - "idle": Library is closed — perform scrape every 60 minutes.
+// - "skip": Library is nonexistent — do not scrape at all.
+type ScrapeStatus = "active" | "idle" | "skip";
+
+export function getScrapeStatus(library: "LL" | "SL" | "LGSC"): ScrapeStatus {
+  const pacificTime = new Date(
+    new Date().toLocaleString("en-US", { timeZone: "America/Los_Angeles" }),
+  );
+  const hour = pacificTime.getHours();
+  const day = pacificTime.getDay();
+
+  if (library === "LL" || library === "SL") {
+    if (day >= 1 && day <= 4 && hour >= 8 && hour < 23) return "active"; // Mon–Thu: 8am–11pm
+    if (day === 5 && hour >= 8 && hour < 18) return "active"; // Fri: 8am–6pm
+    if (day === 6 && hour >= 13 && hour < 17) return "active"; // Sat: 1pm–5pm
+    if (day === 0 && hour >= 13 && hour < 21) return "active"; // Sun: 1pm–9pm
+    return "idle";
+  }
+
+  if (library === "LGSC") {
```

Comment:
continuing on the above, it would help if those comments were reasonably specific to the library being scraped; e.g. does langson have different hours to science library?